src/phongo_bson.c

```diff
@@ -790,37 +791,51 @@ static bool php_phongo_bson_visit_document(const bson_iter_t* iter ARG_UNUSED, c
 	php_phongo_bson_state* parent_state = (php_phongo_bson_state*) data;
 	php_phongo_bson_state  state;
 
-	if (!bson_iter_init(&child, v_document)) {
-		return false;
-	}
-
 	php_phongo_field_path_push(parent_state->field_path, key, PHONGO_FIELD_PATH_ITEM_DOCUMENT);
 
 	PHONGO_BSON_INIT_STATE(state);
 	php_phongo_bson_state_copy_ctor(&state, parent_state);
 
-	array_init(&state.zchild);
-
-	if (bson_iter_visit_all(&child, &php_bson_visitors, &state) || child.err_off) {
-		/* Iteration stopped prematurely due to corruption or a failed
-		 * visitor. Free state.zchild, which we just initialized, and return
-		 * true to stop iteration for our parent context. */
-		zval_ptr_dtor(&state.zchild);
-		php_phongo_bson_state_dtor(&state);
-		return true;
-	}
-
 	/* Check for entries in the fieldPath type map key, and use them to
 	 * override the default ones for this type */
 	php_phongo_handle_field_path_entry_for_compound_type(&state, &state.map.document);
 
+	/* Only traverse BSON document if we're not returning a raw BSON structure */
+	if (state.map.document.type != PHONGO_TYPEMAP_BSON) {
+		if (!bson_iter_init(&child, v_document)) {
+			php_phongo_bson_state_dtor(&state);
+			return false;
+		}
+
+		array_init(&state.zchild);
+
+		if (bson_iter_visit_all(&child, &php_bson_visitors, &state) || child.err_off) {
+			/* Iteration stopped prematurely due to corruption or a failed
+			 * visitor. Free state.zchild, which we just initialized, and return
+			 * true to stop iteration for our parent context. */
+			zval_ptr_dtor(&state.zchild);
+			php_phongo_bson_state_dtor(&state);
+			return true;
+		}
+	}
+
 	/* If php_phongo_bson_visit_binary() finds an ODM class, it should
 	 * supersede a default type map and named document class. */
 	if (state.odm && state.map.document.type == PHONGO_TYPEMAP_NONE) {
 		state.map.document.type = PHONGO_TYPEMAP_CLASS;
 	}
 
 	switch (state.map.document.type) {
+		case PHONGO_TYPEMAP_BSON: {
+			php_phongo_document_t* intern;
+
+			object_init_ex(&state.zchild, php_phongo_document_ce);
```

Comment:
No - I'd like to clarify that in the documentation that as soon as the type map encounters a `bson` type, it will not honour the type map for any objects in its tree. The type map is also not passed on to the `Document` or `ArrayList` instance in question.

Reply:
So we don't loose track of this, you may want to build up the issue description in PHPC-326 with a summary of the implementation details. Alternatively, leave a comment to make note of specific things we'll want to cover in the docs.

Reply:
Created PHPC-2177 to track documentation requirements.